### ISSUE TITLE
Tesseract PDF output corrupted

### ISSUE THREAD

**Issue by myanonymousname** [bytedeco/javacpp-presets#302]
When using JavaCPP Presets for Tesseract (on 64-bit Ubuntu Linux) the pdf output is corrupted and not readable at all for most document viewers.  I'm not getting any "formal" exceptions or errors.

Used code:

        tesseract.TessBaseAPI api = new tesseract.TessBaseAPI();
        if (api.Init(tessdataFolder, tesseractLanguageCode) != 0) {
            throw new IOException("tessdata error");
        }
        
        lept.PIX image = pixRead(inputFile);
        api.SetImage(image);

        tesseract.TessBaseAPISetOutputName(api, fileFolder + "/FinalPDF");
        tesseract.TessBaseAPISetVariable(api, "tessedit_create_pdf", "T");

        tesseract.TessResultRenderer renderer;
        renderer = tesseract.TessPDFRendererCreate(fileFolder + "/FinalPDF", tessdataFolder);
        tesseract.TessResultRendererBeginDocument(renderer, tessdataFolder);
        tesseract.TessBaseAPIProcessPage(api, image, 0, inputFile, null, 0, renderer);
        tesseract.TessResultRendererEndDocument(renderer);

        pixDestroy(image);
        tesseract.TessDeleteResultRenderer(renderer);
        
        api.End();

**Comment by saudet** [comment]
So, is this a bug in Tesseract itself?

**Comment by myanonymousname** [comment]
I don't think so, the same version of Tesseract works fine alone, just not with JavaCPP Presets.

**Comment by saudet** [comment]
Ok, do you have some sample code in C++ that is known to work correctly?

**Comment by myanonymousname** [comment]
here: http://pastebin.com/3XQeScyV

**Comment by saudet** [comment]
Great, thanks! So, have you tried to replace
```java
tesseract.TessBaseAPI api = new tesseract.TessBaseAPI();
```
with the following?
```java
tesseract.TessBaseAPI api = tesseract.TessBaseAPICreate();
```

**Comment by myanonymousname** [comment]
I tried today, but it made no difference.

**Comment by saudet** [comment]
Did you also try to call `TessBaseAPIInit3(api, tessdata, lang)`?

**Comment by myanonymousname** [comment]
Yes, still the same result.

**Comment by saudet** [comment]
Ok, what about if you also call the same functions for all the rest, what happens? You cannot call different functions and expect the same result.

**Comment by myanonymousname** [comment]
I'm not sure what you mean. The pdf file is not influenced by the changes I made, the functions probably aren't source of this problem. 

The code I tried this time was:

        //tesseract.TessBaseAPI api = new tesseract.TessBaseAPI();
        tesseract.TessBaseAPI api = tesseract.TessBaseAPICreate();
        
        //if (api.Init(tessdataFolder, tesseractLanguageCode) != 0) {
        //    throw new IOException("tessdata error");
        //}
        if (tesseract.TessBaseAPIInit3(api, tessdataFolder, tesseractLanguageCode) != 0) {
            throw new IOException("tessdata error");
        }
        
        lept.PIX image = pixRead(inputFile);
        api.SetImage(image);

        tesseract.TessBaseAPISetOutputName(api, fileFolder + "/FinalPDF");
        tesseract.TessBaseAPISetVariable(api, "tessedit_create_pdf", "T");

        tesseract.TessResultRenderer renderer;
        renderer = tesseract.TessPDFRendererCreate(fileFolder + "/FinalPDF", tessdataFolder);
        tesseract.TessResultRendererBeginDocument(renderer, tessdataFolder);
        tesseract.TessBaseAPIProcessPage(api, image, 0, inputFile, null, 0, renderer);
        tesseract.TessResultRendererEndDocument(renderer);

        pixDestroy(image);
        tesseract.TessDeleteResultRenderer(renderer);
        
        api.End();

**Comment by saudet** [comment]
So, what about if you also call `TessBaseAPIDelete(api)` instead of `api.End()` and do not call `TessDeleteResultRenderer(renderer)`?

**Comment by myanonymousname** [comment]
Then the pdf remains damaged the same way and moreover the final file is held by the program until it ends and cannot be used in later procedures. As far as I know, the file is freed only if TessDeleteResultRenderer(renderer) is used.
I tried the same code on Windows, the pdf was damaged too, but another way, it could be read, but could not be used by other tools like PDFBox. The problem is probably not my code, but the way the pdf is generated and is platform dependent.
Is it possible that communication with Tesseract is not done right for some platforms? Is there a code that works for sure for pdf generation using JavaCPP Presets for Tesseract?

**Comment by saudet** [comment]
Yes it's possible that there is a bug in JavaCPP, of course, but could you provide some sample code in C or C++ that works? The one you provided does not work, because it does not call `TessDeleteResultRenderer(renderer)`, among other things. Until I see some working code in C or C++, I maintain that this is a bug in Tesseract and not JavaCPP. Please provide a working example in C or C++!

**Comment by shengc** [comment]
The problem to me though is if I have to open the pdf document right after it is generated by tesseract within the same jvm, it cannot be loaded.

    // I am using PDFBox to load it.
    [error] Exception in thread "main" java.io.IOException: expected='obj' actual=''org.apache.pdfbox.io.PushBackInputStream@28ac3dc3

I made calls to all the following methods before loading the pdf document...

    TessResultRendererEndDocument(renderer)
    renderer.deallocate()
    pixDestroy(image)
    api.End()

However, if the document is first generated and the process exits, then if I run another program to load it, and this time it works.

**Comment by saudet** [comment]
Then that's a problem with Tesseract. Make sure to file a bug report
upstream.

**Comment by saudet** [comment]
Is the same thing still happening with 3.05.01?
http://search.maven.org/#search%7Cga%7C1%7Ctesseract

**Comment by sunjianxy301** [comment]
I have been suffering the same problem (also on ubuntu 16.04LTS). I found the output pdf is only corrupt until the java program exit, in other words, the file become good just after program exit. It seems to me that linux system doesn't flush the output of the pdf renderer until program exit. I saw ariticle saying that the C runtime only guarantee to flush files at program exit. Maybe that's the real cause of this issue. But I don't know how to solve it. So I write this comment seeking for help.

**Comment by saudet** [comment]
@sunjianxy301 The file won't get closed unless we call `TessDeleteResultRenderer()`.
